File: edubooks-api/libros/google_books_service.py
# libros/google_books_service.py
import requests
import logging
from typing import Optional, Dict, Any
from django.conf import settings

logger = logging.getLogger(__name__)

class GoogleBooksService:
    """
    Servicio para integrar con Google Books API y obtener información de libros
    incluyendo imágenes de portada según la categoría/género.
    """
    
    BASE_URL = "https://www.googleapis.com/books/v1/volumes"
    
    # Mapeo de categorías locales a términos de búsqueda de Google Books
    CATEGORIA_MAPPING = {
        'Literatura': 'literature',
        'Programación': 'programming computers',
        'Matemáticas': 'mathematics',
        'Ciencias': 'science',
        'Ingeniería': 'engineering',
        'Negocios': 'business',
        'Historia': 'history',
        'Arte': 'art',
        'Psicología': 'psychology',
        'Filosofía': 'philosophy',
        'Medicina': 'medicine',
        'Derecho': 'law',
        'Ficción': 'fiction',
        'Biografía': 'biography',
        'Educación': 'education',
    }
    
    @classmethod
    def buscar_libro_por_datos(cls, titulo: str, autor: str = None, categoria: str = None) -> Optional[Dict[str, Any]]:
        """
        Busca un libro específico por título y autor en Google Books API.
        
        Args:
            titulo: Título del libro
            autor: Autor del libro (opcional)
            categoria: Categoría del libro (opcional)
            
        Returns:
            Diccionario con información del libro o None si no se encuentra
        """
        try:
            # Construir query de búsqueda
            query_parts = []
            
            if titulo:
                query_parts.append(f'intitle:"{titulo}"')
            
            if autor:
                query_parts.append(f'inauthor:"{autor}"')
                
            if categoria and categoria in cls.CATEGORIA_MAPPING:
                query_parts.append(f'subject:{cls.CATEGORIA_MAPPING[categoria]}')
            
            query = ' '.join(query_parts)
            
            # Realizar petición a Google Books API
            params = {
                'q': query,
                'maxResults': 5,
                'printType': 'books',
                'langRestrict': 'es'  # Priorizar libros en español
            }
            
            response = requests.get(cls.BASE_URL, params=params, timeout=10)
            response.raise_for_status()
            
            data = response.json()
            
            if data.get('totalItems', 0) > 0:
                # Tomar el primer resultado que tenga imagen
                for item in data.get('items', []):
                    volume_info = item.get('volumeInfo', {})
                    image_links = volume_info.get('imageLinks', {})
                    
                    if image_links:
                        return cls._extraer_informacion_libro(item)
                        
            return None
            
        except requests.RequestException as e:
            logger.error(f"Error al consultar Google Books API: {e}")
            return None
        except Exception as e:
            logger.error(f"Error inesperado en búsqueda de libro: {e}")
            return None
    
    @classmethod
    def buscar_por_categoria(cls, categoria: str, limite: int = 10) -> list:
        """
        Busca libros por categoría/género en Google Books API.
        
        Args:
            categoria: Categoría a buscar
            limite: Número máximo de resultados
            
        Returns:
            Lista de diccionarios con información de libros
        """
        try:
            if categoria not in cls.CATEGORIA_MAPPING:
                return []
                
            subject = cls.CATEGORIA_MAPPING[categoria]
            
            params = {
                'q': f'subject:{subject}',
                'maxResults': min(limite, 40),  # Google Books limita a 40
                'printType': 'books',
                'langRestrict': 'es',
                'orderBy': 'relevance'
            }
            
            response = requests.get(cls.BASE_URL, params=params, timeout=10)
            response.raise_for_status()
            
            data = response.json()
            libros = []
            
            for item in data.get('items', []):
                libro_info = cls._extraer_informacion_libro(item)
                if libro_info and libro_info.get('imagen_portada'):
                    libros.append(libro_info)
                    
            return libros
            
        except requests.RequestException as e:
            logger.error(f"Error al buscar por categoría en Google Books API: {e}")
            return []
        except Exception as e:
            logger.error(f"Error inesperado en búsqueda por categoría: {e}")
            return []
    
    @classmethod
    def _extraer_informacion_libro(cls, item: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """
        Extrae información relevante de un item de Google Books API.
        
        Args:
            item: Item de respuesta de Google Books API
            
        Returns:
            Diccionario con información del libro
        """
        try:
            volume_info = item.get('volumeInfo', {})
            image_links = volume_info.get('imageLinks', {})
            
            # Priorizar imágenes de mayor calidad
            imagen_portada = None
            for size in ['extraLarge', 'large', 'medium', 'small', 'thumbnail', 'smallThumbnail']:
                if size in image_links:
                    imagen_portada = image_links[size]
                    # Convertir a HTTPS si es necesario
                    if imagen_portada.startswith('http://'):
                        imagen_portada = imagen_portada.replace('http://', 'https://')
                    break
            
            # Retornar información del libro aunque no tenga imagen
            return {
                'titulo': volume_info.get('title', ''),
                'autor': ', '.join(volume_info.get('authors', [])),
                'editorial': volume_info.get('publisher', ''),
                'año_publicacion': cls._extraer_año_publicacion(volume_info.get('publishedDate', '')),
                'descripcion': volume_info.get('description', ''),
                'imagen_portada': imagen_portada,  # Puede ser None
                'isbn': cls._extraer_isbn(volume_info.get('industryIdentifiers', [])),
                'categorias': volume_info.get('categories', []),
                'google_books_id': item.get('id', ''),
                'paginas': volume_info.get('pageCount', 0)
            }
            
        except Exception as e:
            logger.error(f"Error extrayendo información del libro: {e}")
            return None
    
    @classmethod
    def _extraer_año_publicacion(cls, fecha_publicacion: str) -> Optional[int]:
        """
        Extrae el año de publicación de una fecha en formato string.
        """
        try:
            if fecha_publicacion:
                # Tomar los primeros 4 dígitos como año
                año_str = ''.join(filter(str.isdigit, fecha_publicacion))[:4]
                if len(año_str) == 4:
                    año = int(año_str)
                    if 1000 <= año <= 2030:
                        return año
            return None
        except (ValueError, TypeError):
            return None
    
    @classmethod
    def _extraer_isbn(cls, identifiers: list) -> Optional[str]:
        """
        Extrae el ISBN de la lista de identificadores.
        """
        try:
            for identifier in identifiers:
                if identifier.get('type') in ['ISBN_13', 'ISBN_10']:
                    return identifier.get('identifier')
            return None
        except (TypeError, AttributeError):
            return None
    
    @classmethod
    def buscar_libro_por_isbn(cls, isbn: str) -> Optional[Dict[str, Any]]:
        """
        Busca un libro específico por ISBN.
        
        Args:
            isbn: ISBN del libro
            
        Returns:
            Diccionario con información del libro o None
        """
        try:
            # Limpiar ISBN
            isbn_limpio = isbn.replace('-', '').replace(' ', '')
            
            params = {
                'q': f'isbn:{isbn_limpio}',
                'maxResults': 1,
                'printType': 'books'
            }
            
            response = requests.get(cls.BASE_URL, params=params, timeout=10)
            response.raise_for_status()
            
            data = response.json()
            
            if data.get('totalItems', 0) > 0:
                item = data.get('items', [])[0]
                return cls._extraer_informacion_libro(item)
                
            return None
            
        except Exception as e:
            logger.error(f"Error buscando libro por ISBN {isbn}: {e}")
            return None
    
    @classmethod
    def buscar_libro_por_titulo(cls, titulo: str) -> Optional[Dict[str, Any]]:
        """
        Busca un libro específico por título.
        
        Args:
            titulo: Título del libro
            
        Returns:
            Diccionario con información del libro o None
        """
        try:
            params = {
                'q': f'intitle:"{titulo}"',
                'maxResults': 1,
                'printType': 'books',
                'langRestrict': 'es'
            }
            
            response = requests.get(cls.BASE_URL, params=params, timeout=10)
            response.raise_for_status()
            
            data = response.json()
            
            if data.get('totalItems', 0) > 0:
                item = data.get('items', [])[0]
                return cls._extraer_informacion_libro(item)
                
            return None
            
        except Exception as e:
            logger.error(f"Error buscando libro por título {titulo}: {e}")
            return None

    @classmethod
    def obtener_imagen_por_categoria_y_titulo(cls, categoria: str, titulo: str, autor: str = None) -> Optional[str]:
        """
        Método específico para obtener solo la URL de imagen de portada.
        
        Args:
            categoria: Categoría del libro
            titulo: Título del libro
            autor: Autor del libro (opcional)
            
        Returns:
            URL de la imagen de portada o None
        """
        libro_info = cls.buscar_libro_por_datos(titulo, autor, categoria)
        if libro_info:
            return libro_info.get('imagen_portada')
        
        # Si no encuentra por título específico, buscar por categoría
        libros_categoria = cls.buscar_por_categoria(categoria, 5)
        if libros_categoria:
            # Retornar la primera imagen encontrada
            return libros_categoria[0].get('imagen_portada')
            
        return None

File: edubooks-app/src/app/core/services/google-books.service.ts
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { ApiService } from './api.service';

export interface GoogleBooksImageResponse {
  imagen_portada: string;
  fuente: string;
}

export interface GoogleBooksSearchResponse {
  libros: GoogleBookInfo[];
  total: number;
  categoria: string;
  fuente: string;
}

export interface GoogleBookInfo {
  titulo: string;
  autor: string;
  editorial: string;
  año_publicacion?: number;
  descripcion: string;
  imagen_portada: string;
  isbn?: string;
  categorias: string[];
  google_books_id: string;
  paginas: number;
}

@Injectable({
  providedIn: 'root'
})
export class GoogleBooksService {

  constructor(private apiService: ApiService) {}

  /**
   * Obtener imagen de portada desde Google Books API
   * @param titulo Título del libro
   * @param autor Autor del libro (opcional)
   * @param categoria Categoría del libro (opcional)
   * @returns Observable con la URL de la imagen
   */
  obtenerImagenPortada(titulo: string, autor?: string, categoria?: string): Observable<GoogleBooksImageResponse> {
    const data: any = { titulo };
    
    if (autor) {
      data.autor = autor;
    }
    
    if (categoria) {
      data.categoria = categoria;
    }
    
    return this.apiService.post<GoogleBooksImageResponse>('/google-books/imagen/', data);
  }

  /**
   * Buscar libros por categoría en Google Books API
   * @param categoria Categoría a buscar
   * @param limite Número máximo de resultados (máximo 20)
   * @returns Observable con lista de libros
   */
  buscarLibrosPorCategoria(categoria: string, limite: number = 10): Observable<GoogleBooksSearchResponse> {
    const data = {
      categoria,
      limite: Math.min(limite, 20)
    };
    
    return this.apiService.post<GoogleBooksSearchResponse>('/google-books/buscar/', data);
  }

  /**
   * Buscar libro por ISBN
   * @param isbn ISBN del libro
   * @returns Observable con información del libro
   */
  buscarLibroPorISBN(isbn: string): Observable<GoogleBookInfo | null> {
    const data = { isbn: isbn.replace(/[\s\-]/g, '') };
    
    return this.apiService.post<any>('/google-books/buscar-isbn/', data)
      .pipe(
        map(response => response.libro || null),
        catchError(error => {
          console.warn('Error buscando por ISBN:', error);
          return of(null);
        })
      );
  }

  /**
   * Buscar libro por título
   * @param titulo Título del libro
   * @returns Observable con información del libro
   */
  buscarLibroPorTitulo(titulo: string): Observable<GoogleBookInfo | null> {
    const data = { titulo: titulo.trim() };
    
    return this.apiService.post<any>('/google-books/buscar-titulo/', data)
      .pipe(
        map(response => response.libro || null),
        catchError(error => {
          console.warn('Error buscando por título:', error);
          return of(null);
        })
      );
  }

  /**
   * Obtener imagen automáticamente basada en los datos del libro
   * @param libroData Datos del libro
   * @returns Observable con la URL de la imagen o null
   */
  obtenerImagenAutomatica(libroData: { titulo: string; autor?: string; categoria?: string }): Observable<string | null> {
    return new Observable(observer => {
      this.obtenerImagenPortada(libroData.titulo, libroData.autor, libroData.categoria)
        .subscribe({
          next: (response) => {
            observer.next(response.imagen_portada);
            observer.complete();
          },
          error: (error) => {
            console.warn('No se pudo obtener imagen de Google Books:', error);
            observer.next(null);
            observer.complete();
          }
        });
    });
  }

  /**
   * Validar si una URL de imagen es válida (simplificado para evitar CORS)
   * @param url URL de la imagen
   * @returns Promise<boolean>
   */
  async validarImagenUrl(url: string): Promise<boolean> {
    // Validación básica de URL sin petición HTTP para evitar CORS
    try {
      const urlObj = new URL(url);
      return urlObj.protocol === 'https:' && 
             (urlObj.hostname.includes('googleusercontent.com') || 
              urlObj.hostname.includes('books.google.com'));
    } catch {
      return false;
    }
  }

  /**
   * Obtener imagen con fallback a imagen por defecto
   * @param titulo Título del libro
   * @param autor Autor del libro
   * @param categoria Categoría del libro
   * @param imagenPorDefecto URL de imagen por defecto
   * @returns Promise<string> URL de la imagen final
   */
  async obtenerImagenConFallback(
    titulo: string, 
    autor: string, 
    categoria: string, 
    imagenPorDefecto: string
  ): Promise<string> {
    try {
      const response = await this.obtenerImagenPortada(titulo, autor, categoria).toPromise();
      
      if (response?.imagen_portada) {
        // Validar formato básico de URL sin petición HTTP
        const esValida = await this.validarImagenUrl(response.imagen_portada);
        if (esValida) {
          return response.imagen_portada;
        }
      }
    } catch (error) {
      console.warn('Error obteniendo imagen de Google Books:', error);
    }
    
    // Retornar imagen por defecto si no se encuentra o hay error
    return imagenPorDefecto;
  }
}